### PR TITLE
MM-56540 - Live-captions support

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- At the moment this is pretty WIP, making a PR to get feedback on the general approach.
- I am keeping the debug statements (starting with `<><>`) which give a lot of useful insight into what's happening in the transcriber.
- I've left Todos to show what I'm still planning to do, and liberal comments to help explain why things are the way they are.
- I'm open to changing anything and everything, so please suggest anything that comes to mind.

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-56540

### CODE REVIEW

**Comment by cpoile** [comment]
~~Looks like the lint is failing on the C types -- I'm wondering if there's a way to make this work on both linux and darwin?~~

~~Edit: working on the fix, now lints, I'll get it green soon.~~

Fixed.

File: cmd/transcriber/call/transcriber.go
Comment (streamer45) [inline cmd/transcriber/call/transcriber.go]: Any reason for not calling this inside the existing `t.done()` method?

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: This probably belongs in either the `rtcd` client or https://github.com/mattermost/mattermost-plugin-calls/tree/main/server/public . At some point I'd like to unify the two.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Can we use the `Ch` suffix to be super explicit on what `ret` is when used elsewhere?

File: cmd/transcriber/call/tracks.go
Comment (streamer45) [inline cmd/transcriber/call/tracks.go:119]: I think there's a chance to decouple further if we run this logic in its own goroutine. Two things I'd like to see:

- Avoid decoding as part of the reading loop since it's a blocking operation. We should pass the data to a channel and deal with the rest asynchronously.
- Have a config option/flag to entirely disable live captions and only run the basic post-call transcriber. This is important since it will allow our system admins some flexibility especially since there are potentially significant performance considerations to be made.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Not a huge deal but I always found it to be safe/more idiomatic to call `Add` prior to spawning the goroutine rather than inside.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Ideally yes but not a big deal to be honest as Transcribe should return eventually.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Can we use the method name to identify the log?

**Review by streamer45 — COMMENTED** [review]
Looks pretty good 💪 . Gave it a very quick pass to potentially unblock further improvements. Still need to go through the meat (`processLiveCaptionsForTrack`).

My only "blocking" thought would be the need for a config setting to control whether we do live captions. This way we can have a related setting on the plugin side to globally enable/disable this functionality.

File: cmd/transcriber/call/tracks.go
Comment (cpoile) [reply cmd/transcriber/call/tracks.go:119]: Good suggestions, thank you.
- Done, and then I also decoupled the reading packets/decoding into PCM stage from the VAD/transcription phase. This removes the dependency between those steps (the `window` slice), and means we don't have to worry about a pathological case where the server just can't transcribe fast enough. Before, that could have led to blocking the packet reading goroutine, which would fill up the payload buffer, which would have led to blocking the reading loop. Now it won't -- all that will happen is we will add more to the `toBeTranslated` slice, which is fine. We may drop untranslated voice, but that's what we should do anyway (if the server is being crushed).
- I've added config options (a couple to the system console, and one in the env vars -- see the plugin changes, mostly: https://github.com/mattermost/mattermost-plugin-calls/pull/633/commits/477002b82434eeb5563d1959aec273735f7d8a81).

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: Yep, good idea.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: ~~After thinking it over, I think it belongs here -- it isn't used by anyone else. The `rtcd` client doesn't know or care what the ws msg is, and we inspect the ws msg in the `calls` package by field name, not by type (that would be nice to change in the future).~~

File: cmd/transcriber/call/transcriber.go
Comment (cpoile) [reply cmd/transcriber/call/transcriber.go]: Huh, I think that might have changed -- before it wasn't working (causing a deadlock), but now it does. Thanks.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: Thanks!

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: Nope, you were right, I put it in public: https://github.com/mattermost/mattermost-plugin-calls/pull/633/commits/9055277fda92c143acfd66eee72d2a63021c327e

**Review by jupenur — DISMISSED** [review]

**Comment by cpoile** [comment]
@jupenur We should schedule this for full sec review when your time permits. I will be making algorithmic changes over the next few days, but the underlying architecture (using the same external code paths that the after-call-transcriber uses) will remain the same. 
Please let me know if there's anything that you need for the review. Thank you.

**Comment by jupenur** [comment]
@cpoile there was nothing particularly risky here. I can do another quick pass some time next week -- it shouldn't take me more than a few hours tops. Just tag me again once all the changes are in.

**Comment by cpoile** [comment]
Changes are in, @jupenur. Thank you for taking the time.

**Review by jupenur — DISMISSED** [review]
Still looking pretty solid from my PoV. Thanks @cpoile!

File: cmd/transcriber/call/transcriber.go
Comment (streamer45) [inline cmd/transcriber/call/transcriber.go]: What's the use of `msg`? Surprised it's not failing on lint check.

File: cmd/transcriber/call/transcriber.go
Comment (streamer45) [inline cmd/transcriber/call/transcriber.go]: Is this really needed?

File: cmd/transcriber/config/config.go
Comment (streamer45) [inline cmd/transcriber/config/config.go]: This is an important change as it would make defaults much less performant even on high tier systems. Wondering if we could have different settings based on whether live captions are enabled, in which case it makes sense to go with the more conservative value.

File: cmd/transcriber/config/config.go
Comment (streamer45) [inline cmd/transcriber/config/config.go:263]: From your testing, do you think it's still valuable to expose this?

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Is this still the case? Have we tested with the larger `1536`?

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Could we suffix `pktPayloads`? For consistency if anything.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Not in love with the interface here. If we are reading we should be returning something. Keeping it as pure side effect can be slightly confusing.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Just pointing out that this approach makes us even more reliant on speech detector's quality and reliability.

File: cmd/transcriber/call/tracks.go
Comment (streamer45) [inline cmd/transcriber/call/tracks.go]: Wondering if we could use the single `pktPayloadCh` to track whether we are done or not.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go:309]: Not too convinced by this. You can say a lot in one second. I'd rather have artifacts than data loss if I could choose.

**Review by streamer45 — COMMENTED** [review]
Still a lot of complexity in here but if it works it works I suppose :)

Left some candid feedback, nothing truly blocking though.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: How hard would it be to come up with something dynamic? Like autodetecting the first few seconds and then use the resulting lang code throughout, regardless of which track it's coming from. I think it's fair to assume same call == same language.

Otherwise I feel we may need to export it so that it can at least be controlled by admins.

File: cmd/transcriber/call/transcriber.go
Comment (cpoile) [reply cmd/transcriber/call/transcriber.go]: I have no idea how that snuck in there, thanks.

File: cmd/transcriber/call/transcriber.go
Comment (cpoile) [reply cmd/transcriber/call/transcriber.go]: I guess not totally needed -- was using it during development. Will remove.

File: cmd/transcriber/config/config.go
Comment (cpoile) [reply cmd/transcriber/config/config.go]: Yep, that sounds fine.

File: cmd/transcriber/config/config.go
Comment (cpoile) [reply cmd/transcriber/config/config.go:263]: Imo, yes. I would say keep it in there as an option.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: I did test just now, I don't think I can tell much of a difference, so I'd like to stay with 512 since it's what silero-vad seems to have been trained with (its examples and quality metrics pages).

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: Agreed, fixed. Fixed the `removeShortSpeeches` too so we're always returning the result.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: I suppose, but after a lot of testing it's a significant decrease in garbage output, so it definitely helps even with a spotty VAD.

File: cmd/transcriber/call/tracks.go
Comment (cpoile) [reply cmd/transcriber/call/tracks.go]: Yep that works, good suggestion, thanks.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go:309]: We do need some smoothing here, or else you get more hallucinations. One of the Silero scientists states that most meaningful speech is longer than 250ms (https://thegradient.pub/one-voice-detector-to-rule-them-all/) so I did some experiments around that number. I tended to see more hallucinations when the minSpeechLength was 150ms, but fewer at 210 and 240, then I couldn't tell a huge difference above that. So I think 210ms (actually 330 bc of padding) seems like a good barrier. This means that if we have [`silence` (>=150ms), `speech` (<=210ms + 120ms padding), `silence` (>=150ms)], we can safely remove the speech.
So yes, we're accepting a bit of data loss, but the result is fewer hallucinations and garbage output. I think that's a worthwhile tradeoff considering the type of speech we'll be losing at that level will be short "um" and "ah"s. Longer "um"s and "ah"s are still detected.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: I'm hesitant to try to detect it, considering the trouble we've had with that in the past, and we'd need rules (what confidence, where in the meeting do we detect, for how long, etc.), and then there's the issue of meetings that start with one language and go to the main language later. If we're going to do a single set language, probably we should make it per-call configurable, or system-level (perhaps in the system-console).
Needs UX, so made a ticket: https://mattermost.atlassian.net/browse/MM-57408

Comment (streamer45) [reply cmd/transcriber/call/live_captions.go]: Yes, I was assuming system console (global setting) which wouldn't need much UX work, either a dropdown or a simple text field expecting the country code. I'd like to avoid the hard-coded `en` since we already support autodetection for post call transcripts.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [reply cmd/transcriber/call/live_captions.go:309]: Thanks for the thorough explanation. ~300ms sounds like a much more reasonable threshold.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: Cool, added the config setting -- I'll add the system console setting in the main PR.

Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: Done in https://github.com/mattermost/mattermost-plugin-calls/pull/633/commits/80602b99a142f3b886d9b929db09f025d2d8800f

File: cmd/transcriber/call/transcriber.go
Comment (streamer45) [inline cmd/transcriber/call/transcriber.go]: Could we rename them slightly to be more explicit as these refer to the transcriber pool for live captions. Or add a comment.

Comment (streamer45) [inline cmd/transcriber/call/transcriber.go]: Is there a reason we couldn't use the existing global `doneCh`?

File: go.mod
Comment (streamer45) [inline go.mod]: We should update this to point to master.

File: go.mod
Comment (streamer45) [inline go.mod]: Same, once merged.

File: cmd/transcriber/call/transcriber.go
Comment (streamer45) [reply cmd/transcriber/call/transcriber.go]: Yeah, we can infer if they are disabled. It's also a less interesting case.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go]: Is it expected/safe not calling `t.transcriberWg.Done()` in this case?

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go:317]: Wondering if it'd be possible to do this in place instead of allocating every time.

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [inline cmd/transcriber/call/live_captions.go:235]: Could we add a `continue` here and avoid the loop label altogether?

**Review by streamer45 — COMMENTED** [review]
I promise these are the last comments :)

File: cmd/transcriber/call/transcriber.go
Comment (cpoile) [reply cmd/transcriber/call/transcriber.go]: Changed to captionsPool, thanks.

Comment (cpoile) [reply cmd/transcriber/call/transcriber.go]: Yes, the captionsPool needs to be closed, then transcriptions happen, then global `doneCh` is called. I could rearrange that, but then logically it's not as clean.

File: go.mod
Comment (cpoile) [reply go.mod]: Can we cut a new version and point to that?

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go]: Right, should do that here too, thanks.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go:317]: I had it like that, but then went this way because I didn't want to permanently change the window for subsequent ticks. I'm not worried about the small memory cost, when you compare it to the cost of the transcriber.

File: cmd/transcriber/call/live_captions.go
Comment (cpoile) [reply cmd/transcriber/call/live_captions.go:235]: Unless I'm misunderstanding what you mean, adding a continue here is the same as not having a continue, and you would still need to break with the label to escape the select (which captures the break).

File: go.mod
Comment (streamer45) [reply go.mod]: Yes, it's practically the same though :)

File: cmd/transcriber/call/live_captions.go
Comment (streamer45) [reply cmd/transcriber/call/live_captions.go:235]: @cpoile It would be `continue` in the ticker case and a `break` in the for loop after the select block.

Comment (cpoile) [reply cmd/transcriber/call/live_captions.go:235]: @streamer45 Ah, I _was_ misunderstanding what you were getting at. Yep, that's better.

File: go.mod
Comment (streamer45) [reply go.mod]: https://github.com/mattermost/rtcd/tree/v0.14.0

**Review by streamer45 — DISMISSED** [review]

**Review by streamer45 — APPROVED** [review]

**Pull request merged**
